File: python/examples/extranonce_allocator_example.py
"""
Stratum v2 Extranonce Allocator Example

This example demonstrates how to leverage Sv2ExtranonceAllocator.

A Mining Server could have multiple clients, and each client could have multiple channels.
In Sv2, the extranonce prefix is what lets the Mining Server guarantee unique search space for
different channels.

A Mining Server should use a single Sv2ExtranonceAllocator for both Standard and Extended
Channels. This keeps standard and extended channel prefixes in one allocation namespace.
"""

from sv2 import Sv2ExtranonceAllocator, sv2_extranonce_bytes_needed


# Size of the static identifier for this pool server, placed at the start of the pool's
# extranonce allocation. One byte covers up to 256 distinct pool servers.
POOL_SERVER_BYTES = 1
# Maximum number of concurrent channels the pool can allocate. Determines
# `POOL_LOCAL_PREFIX_BYTES` via `sv2_extranonce_bytes_needed`. The internal
# allocation bitmap uses `POOL_MAX_CHANNELS / 8` bytes of RAM.
POOL_MAX_CHANNELS = 16_777_216
# Bytes consumed by the per-channel `local_index`. Derived from
# `POOL_MAX_CHANNELS` so the two stay in sync.
POOL_LOCAL_PREFIX_BYTES = sv2_extranonce_bytes_needed(POOL_MAX_CHANNELS)
POOL_ALLOCATION_BYTES = POOL_SERVER_BYTES + POOL_LOCAL_PREFIX_BYTES
CLIENT_SEARCH_SPACE_BYTES = 16
FULL_EXTRANONCE_SIZE = POOL_ALLOCATION_BYTES + CLIENT_SEARCH_SPACE_BYTES


class MiningServer:
    extranonce_allocator: Sv2ExtranonceAllocator

    def __init__(self, server_id: int):
        local_prefix_bytes = server_id.to_bytes(POOL_SERVER_BYTES, "big")
        self.extranonce_allocator = Sv2ExtranonceAllocator(
            local_prefix_bytes=local_prefix_bytes,
            total_extranonce_len=FULL_EXTRANONCE_SIZE,
            max_channels=POOL_MAX_CHANNELS,
        )


def main():
    """Main demonstration function."""
    print("Stratum v2 Extranonce Allocator Example")
    print("=" * 50)

    mining_server_a = MiningServer(0)
    mining_server_b = MiningServer(1)

    standard_a_1 = mining_server_a.extranonce_allocator.allocate_standard()
    standard_a_2 = mining_server_a.extranonce_allocator.allocate_standard()
    print(f"Standard Channel 1 on Mining Server A: {standard_a_1.as_bytes().hex()}")
    print(f"Standard Channel 2 on Mining Server A: {standard_a_2.as_bytes().hex()}")

    standard_b_1 = mining_server_b.extranonce_allocator.allocate_standard()
    standard_b_2 = mining_server_b.extranonce_allocator.allocate_standard()
    print(f"Standard Channel 1 on Mining Server B: {standard_b_1.as_bytes().hex()}")
    print(f"Standard Channel 2 on Mining Server B: {standard_b_2.as_bytes().hex()}")

    extended_a_1 = mining_server_a.extranonce_allocator.allocate_extended(
        min_rollable_size=CLIENT_SEARCH_SPACE_BYTES
    )
    extended_a_2 = mining_server_a.extranonce_allocator.allocate_extended(
        min_rollable_size=CLIENT_SEARCH_SPACE_BYTES
    )
    extended_b_1 = mining_server_b.extranonce_allocator.allocate_extended(
        min_rollable_size=CLIENT_SEARCH_SPACE_BYTES
    )
    extended_b_2 = mining_server_b.extranonce_allocator.allocate_extended(
        min_rollable_size=CLIENT_SEARCH_SPACE_BYTES
    )
    print(f"Extended Channel 1 on Mining Server A: {extended_a_1.as_bytes().hex()}")
    print(f"Extended Channel 2 on Mining Server A: {extended_a_2.as_bytes().hex()}")
    print(f"Extended Channel 1 on Mining Server B: {extended_b_1.as_bytes().hex()}")
    print(f"Extended Channel 2 on Mining Server B: {extended_b_2.as_bytes().hex()}")

    try:
        mining_server_a.extranonce_allocator.allocate_extended(
            min_rollable_size=CLIENT_SEARCH_SPACE_BYTES + 1
        )
    except Exception:
        print(
            "Client wants to roll more bytes than this allocator allows. "
            "Return an OpenMiningChannel.Error."
        )


if __name__ == "__main__":
    main()

File: cpp/examples/extranonce_allocator_example.cpp
/*
 * Stratum v2 Extranonce Allocator Example
 *
 * This example demonstrates how to leverage Sv2ExtranonceAllocator.
 *
 * A Mining Server could have multiple clients, and each client could have multiple channels.
 * In Sv2, the extranonce prefix is what lets the Mining Server guarantee unique search space for
 * different channels.
 *
 * A Mining Server should use a single Sv2ExtranonceAllocator for both Standard and Extended
 * Channels. This keeps standard and extended channel prefixes in one allocation namespace.
 */

#include <cstdint>
#include <exception>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include "example_utils.hpp"
#include "sv2.hpp"

namespace {

constexpr std::uint32_t POOL_SERVER_BYTES = 1;
constexpr std::uint32_t POOL_MAX_CHANNELS = 16'777'216;
constexpr std::uint32_t CLIENT_SEARCH_SPACE_BYTES = 16;

class MiningServer {
public:
    explicit MiningServer(std::uint8_t server_id)
        : extranonce_allocator(sv2::Sv2ExtranonceAllocator::init(
              std::vector<std::uint8_t>{server_id},
              POOL_SERVER_BYTES + sv2::sv2_extranonce_bytes_needed(POOL_MAX_CHANNELS) + CLIENT_SEARCH_SPACE_BYTES,
              POOL_MAX_CHANNELS
          )) {
        if (extranonce_allocator == nullptr) {
            throw std::runtime_error("Sv2ExtranonceAllocator::init returned null");
        }
    }

    std::shared_ptr<sv2::Sv2ExtranonceAllocator> extranonce_allocator;
};

void print_prefix(
    const std::string &label,
    const std::shared_ptr<sv2::Sv2ExtranoncePrefix> &prefix
) {
    if (prefix == nullptr) {
        throw std::runtime_error(label + " allocation returned null");
    }

    std::cout
        << label << ": "
        << sv2cpp::example::hex(prefix->as_bytes())
        << " (" << prefix->len() << " bytes)"
        << std::endl;
}

} // namespace

int main() {
    try {
        std::cout << "Stratum v2 Extranonce Allocator Example" << std::endl;
        std::cout << "==================================================" << std::endl;

        std::cout << "Pool server bytes: " << POOL_SERVER_BYTES << std::endl;
        std::cout << "Pool max channels: " << POOL_MAX_CHANNELS << std::endl;
        std::cout
            << "Pool local index bytes: "
            << sv2::sv2_extranonce_bytes_needed(POOL_MAX_CHANNELS)
            << std::endl;
        std::cout
            << "Full extranonce size: "
            << POOL_SERVER_BYTES + sv2::sv2_extranonce_bytes_needed(POOL_MAX_CHANNELS) + CLIENT_SEARCH_SPACE_BYTES
            << std::endl;
        std::cout
            << "Client search space bytes: "
            << CLIENT_SEARCH_SPACE_BYTES
            << std::endl;
        std::cout << std::endl;

        // different mining servers have unique server bytes
        // this guarantees no extranonce collisions between servers
        MiningServer mining_server_a(0);
        MiningServer mining_server_b(1);

        const auto standard_a_1 = mining_server_a.extranonce_allocator->allocate_standard();
        const auto standard_a_2 = mining_server_a.extranonce_allocator->allocate_standard();
        const auto standard_b_1 = mining_server_b.extranonce_allocator->allocate_standard();
        const auto standard_b_2 = mining_server_b.extranonce_allocator->allocate_standard();

        print_prefix("Standard Channel 1 on Mining Server A", standard_a_1);
        print_prefix("Standard Channel 2 on Mining Server A", standard_a_2);
        print_prefix("Standard Channel 1 on Mining Server B", standard_b_1);
        print_prefix("Standard Channel 2 on Mining Server B", standard_b_2);

        std::cout << std::endl;

        // clients can potentially ask for less than CLIENT_SEARCH_SPACE_BYTES
        const auto extended_a_1 = mining_server_a.extranonce_allocator->allocate_extended(CLIENT_SEARCH_SPACE_BYTES - 8);
        const auto extended_a_2 = mining_server_a.extranonce_allocator->allocate_extended(CLIENT_SEARCH_SPACE_BYTES - 4);
        const auto extended_b_1 = mining_server_b.extranonce_allocator->allocate_extended(CLIENT_SEARCH_SPACE_BYTES - 2);
        const auto extended_b_2 = mining_server_b.extranonce_allocator->allocate_extended(CLIENT_SEARCH_SPACE_BYTES);

        print_prefix("Extended Channel 1 on Mining Server A", extended_a_1);
        print_prefix("Extended Channel 2 on Mining Server A", extended_a_2);
        print_prefix("Extended Channel 1 on Mining Server B", extended_b_1);
        print_prefix("Extended Channel 2 on Mining Server B", extended_b_2);

        std::cout << std::endl;
        std::cout
            << "Mining Server A allocated "
            << mining_server_a.extranonce_allocator->allocated_count()
            << " prefixes"
            << std::endl;
        std::cout
            << "Mining Server B allocated "
            << mining_server_b.extranonce_allocator->allocated_count()
            << " prefixes"
            << std::endl;

        std::cout << std::endl;
        std::cout << "--- Invalid request demonstration ---" << std::endl;

        // clients can't ask for more than CLIENT_SEARCH_SPACE_BYTES
        try {
            (void)mining_server_a.extranonce_allocator->allocate_extended(CLIENT_SEARCH_SPACE_BYTES + 1);
            std::cout
                << "Unexpectedly accepted an oversized rollable extranonce request"
                << std::endl;
        } catch (const sv2::Sv2ExtranonceAllocatorError &error) {
            std::cout
                << "Client wants to roll more bytes than this allocator allows. "
                << "Return an OpenMiningChannel.Error."
                << std::endl;
            (void)error;
        }

        std::cout << std::endl;
        std::cout << "Example completed successfully" << std::endl;
        return 0;
    } catch (const std::exception &error) {
        std::cerr << "Example failed: " << error.what() << std::endl;
        return 1;
    } catch (...) {
        std::cerr << "Example failed with an unknown error" << std::endl;
        return 1;
    }
}